spec/ddtrace/contrib/grape/tracer_spec.rb

```diff
@@ -353,12 +374,124 @@
+          expect(run_span.get_tag(Datadog::Ext::HTTP::URL)).to eq('/widgets')
+
+          expect(run_span.get_tag(Datadog::Contrib::Grape::Ext::TAG_ROUTE_PATH)).to eq('/widgets')
+          expect(run_span.get_tag(Datadog::Contrib::Grape::Ext::TAG_ROUTE_METHOD)).to eq('GET')
+        end
+      end
+
+      context 'post method' do
+        subject(:response) { post '/widgets' }
+
+        it_behaves_like 'measured span for integration', true do
+          before { expect(response.status).to eq(201) }
+        end
+
+        it_behaves_like 'analytics for integration', ignore_global_flag: false do
+          let(:analytics_enabled_var) { Datadog::Contrib::Grape::Ext::ENV_ANALYTICS_ENABLED }
+          let(:analytics_sample_rate_var) { Datadog::Contrib::Grape::Ext::ENV_ANALYTICS_SAMPLE_RATE }
+          before { expect(response.status).to eq(201) }
+        end
+
+        it 'should trace the endpoint body' do
+          expect(response.status).to eq(201)
+          expect(response.body).to eq('{}')
+          expect(spans.length).to eq(2)
+
+          expect(render_span.name).to eq('grape.endpoint_render')
+          expect(render_span.span_type).to eq('template')
+          expect(render_span.service).to eq('grape')
+          expect(render_span.resource).to eq('grape.endpoint_render')
+          expect(render_span).to_not have_error
+          expect(render_span.parent).to eq(run_span)
+
+          expect(run_span.name).to eq('grape.endpoint_run')
+          expect(run_span.span_type).to eq('web')
+          expect(run_span.service).to eq('grape')
+          expect(run_span.resource).to eq('TestingAPI POST /widgets')
+          expect(run_span).to_not have_error
+          expect(run_span.parent).to be_nil
+
+          expect(run_span.get_tag(Datadog::Ext::HTTP::METHOD)).to eq('POST')
+          expect(run_span.get_tag(Datadog::Ext::HTTP::URL)).to eq('/widgets')
+
+          expect(run_span.get_tag(Datadog::Contrib::Grape::Ext::TAG_ROUTE_PATH)).to eq('/widgets')
+          expect(run_span.get_tag(Datadog::Contrib::Grape::Ext::TAG_ROUTE_METHOD)).to eq('POST')
+        end
+      end
+
+      context 'deeply nested' do
+        subject(:response) { get '/nested/widgets' }
+
+        it 'should trace the endpoint body' do
+          is_expected.to be_ok
+          expect(response.body).to eq('[]')
+          expect(spans.length).to eq(2)
+
+          expect(render_span.name).to eq('grape.endpoint_render')
+          expect(render_span.span_type).to eq('template')
+          expect(render_span.service).to eq('grape')
+          expect(render_span.resource).to eq('grape.endpoint_render')
+          expect(render_span).to_not have_error
+          expect(render_span.parent).to eq(run_span)
+
+          expect(run_span.name).to eq('grape.endpoint_run')
+          expect(run_span.span_type).to eq('web')
+          expect(run_span.service).to eq('grape')
+          expect(run_span.resource).to eq('TestingAPI GET /nested/widgets')
```

Comment:
Similar to the suggestion above, but displaying how it would work for this nested case.
```suggestion
          expect(run_span.resource).to eq('TestingAPI::nested::widgets#get')
```

Reply:
@marcotc Thanks for the thoughtful feedback! Before looking at some of your specific suggestions, I wanted to understand a bit more about your goals for the resource naming pattern. Your naming recommendation appears to follow relatively closely to the Rails tracer which has an additional routing layer.

In my opinion, Grape  has much more knowledge of the actual request path structure than a Rails controller does which puts Grape in a category much closer to Sinatra in practice. For that reason, I prefer to structure the naming similar to Sinatra.

Reply:
I guess the main question here is: would you rather visually identify your traces by the Ruby method that handled them (`TestingAPI::nested::widgets#get`), or by the HTTP route they took (`TestingAPI GET /nested/widgets`)?

No information is lost in either case, as all information still collected as tags. But my case here is: how would you like to read these, when looking at an overview page with dozens of results?

In my personal experience (which is definitely anecdotal), I see the HTTP method and path as a lower-level concern _when browsing the service page_. While seeing `TestingAPI::nested::widgets#get` would lead me to directly ask: "did our team touch the `get` method in the `widget` namespace recently?".

That being said, I do agree that Grape sits in between Sinatra and Rails when it comes to the level of abstraction atop HTTP (Sinatra being lower level, Rails being higher level).

At the end of day, your experience as a current production user of Grape is the strongest signal here. If believe that `TestingAPI GET /nested/widgets` gives you a more immediate signal to where in the application this request is being handled, I'm more than happy to go with the current implementation.

My suggestion was trying to along the lines of: "Is the proposed representation the cleanest, most immediately identifiable representation we can think of, while ensuring that it uniquely identifies an endpoint?".

Reply:
I followed up with our internal teams and the overall consensus is that using a path similar to the Sinatra trace would be preferred. In general, the paths are going to be more generic and understandable for stakeholders that are less plugged into the actual Grape architecture (e.g. engineering leadership and SRE) while still being understandable for developers working directly w/ Grape. Therefor, I'd prefer to stick with the current proposed naming pattern.